index.rst

```diff
@@ -1,61 +1,270 @@
+============
 
-  Sections
-  ========
+"Middleware" is the term used to describe software applications used for communication in software systems.
+These technologies became popular with the advent of *distributed systems*, which come as a solution to the problem of parallel computation.
+As software and hardware systems became more complex, it becomes impractical to develop and execute them in a single process and, eventually, a single node.
+With that, software evolved from *monolith* applications, where a single program executes in a single process or node, to *distributed* applications, where the system is divided into a number of smaller applications; each running on their own process or node.
+For these applications to work together coherently they must be able to communicate with each other, thus giving origin to middleware technologies.
 
-  and
+How a *distributed system* is broken down into smaller pieces is heavily dependent upon the problem.
+Some systems are only broken down into a small number of components each still in charge of large contexts, others are broken down into many small applications that are in charge only of small simple tasks.
+The latter has gained substantial popularity recently and is commonly referred to as *microservices*.
+These systems are behind many of popular large services in use today like Google and Amazon. 
 
-  Subsections
-  -----------
+The architecture of *distributed systems* can take many shapes and forms.
+For instance, some systems are designed to emulate *monolithic* applications.
+The application is composed of a number of smaller applications but there is a hierarchical organization, with components at the top, in charge of communicating and operating components at the bottom.
+The advantage of these systems is that they are easier to understand and to maintain.
+Since each component is isolated from the rest of the system, and only communicates with a components on the top and bottom of the hierarchical chain, adding new components is relatively easy and have minimal impact in the system.
+The disadvantage is that the system is more vulnerable to outages, if a component on the top of the hierarchical chain becomes unavailable those components at the bottom also become unavailable.
 
-  and
+A popular alternative to this architecture are *event-driven reactive systems*.
+In these systems each component is designed as an independent entity that reacts to input data, be it from other components or external data services.
+Since these system are designed with separation of concerns in mind, each component must be able to act independently, *reactive systems* are usually extremely resilient.
+If one component becomes unavailable, the others are expected to continue to operate, taking precautious to deal with the missing agent.
+At the same time, it can become quite burdensome to update and grow these systems as their complexity increases exponentially with the number of different components.
```

Comment:
I am still very concerned about this section. When I read this I think you are saying that we would be better off with a hierarchical system. That may be true. It may be that if we were starting over we would choose a hierarchical model. But I think it is not a practical choice now because SAL is not designed that way.

Also I am not convinced these two architectures are orthogonal. Hierarchy is a matter of degree. At one end we have shallow and wide designs (many systems at the top level, with only one or a few levels of hierarchy below that), and at the other end we have deep and narrow designs (few systems at the top level with many levels beneath them). That said, our system has no hierarchy at all -- the extreme case of shallow and wide.

It might be better off to omit the discussion of hierarchical systems. If you want to keep it then please clarify that we do not consider it a viable option for Rubin and explain why.

Reply:
Ok I think I see your point. I rephrased it..